### PR TITLE
Adjust ds microblock gas limit after view change

### PR DESCRIPTION
**Pull request by kaikawaliu**
## Description

Exponentially reduce the gas limit for ds microblock if vc counter increased.
And also put microblock gas limit under consensus

## Backward Compatibility

- [ ] This is not a breaking change
- [x] This is a breaking change

### Implementation

- [x] **ready for review**

### Integration Test (Core Team)

- [ ] local machine test

- [x] small-scale cloud test (with vc test checking)

### CODE REVIEW

File: src/libNode/Node.h [hunk 1]
```diff
@@ -294,7 +294,8 @@ class Node : public Executable {
   bool CheckMicroBlockVersion();
   bool CheckMicroBlockshardId();
   bool CheckMicroBlockTimestamp();
-  bool CheckMicroBlockHashes(bytes& errorMsg);
+  bool CheckMicroBlockHashes(bytes& errorMsg,
+                             const uint64_t& microblock_gas_limit);
```
Comment (ShengguangXiao) [inline src/libNode/Node.h]: Seperate the check gas limit to another function?

File: src/libDirectoryService/DirectoryService.h [hunk 2]
```diff
@@ -527,6 +525,9 @@ class DirectoryService : public Executable {
   /// Force multicast when sending block to shard
   std::atomic<bool> m_forceMulticast{};
 
+  /// microblock_gas_limit to be adjusted due to vc
+  uint64_t m_microblock_gas_limit = MICROBLOCK_GAS_LIMIT;
```
Comment (ShengguangXiao) [inline src/libDirectoryService/DirectoryService.h]: m_microBlockGasLimit

File: src/libDirectoryService/FinalBlockPreProcessing.cpp [hunk 3]
```diff
@@ -1154,6 +1157,45 @@ void DirectoryService::RunConsensusOnFinalBlock() {
     LOG_GENERAL(INFO, "RunConsensusOnFinalBlock ");
     PrepareRunConsensusOnFinalBlockNormal();
 
+    m_microblock_gas_limit = MICROBLOCK_GAS_LIMIT;
+    if (!m_mediator.GetIsVacuousEpoch() &&
+        ((m_mediator.m_dsBlockChain.GetLastBlock()
```
Comment (bb111189) [inline src/libDirectoryService/FinalBlockPreProcessing.cpp]: add a comment on this piece on code

File: src/libDirectoryService/FinalBlockPreProcessing.cpp [hunk 4]
```diff
@@ -1154,6 +1157,45 @@ void DirectoryService::RunConsensusOnFinalBlock() {
     LOG_GENERAL(INFO, "RunConsensusOnFinalBlock ");
     PrepareRunConsensusOnFinalBlockNormal();
 
+    m_microblock_gas_limit = MICROBLOCK_GAS_LIMIT;
```
Comment (bb111189) [inline src/libDirectoryService/FinalBlockPreProcessing.cpp]: Add a comment above to talk about the logic of exponential backoff

File: src/libNode/Node.h (hunk 1 above)
Comment (kaikawaliu) [reply src/libNode/Node.h]: Fixed

File: src/libDirectoryService/DirectoryService.h (hunk 2 above)
Comment (kaikawaliu) [reply src/libDirectoryService/DirectoryService.h]: Changed

File: src/libDirectoryService/FinalBlockPreProcessing.cpp (hunk 3 above)
Comment (kaikawaliu) [reply src/libDirectoryService/FinalBlockPreProcessing.cpp]: Fixed

File: src/libDirectoryService/FinalBlockPreProcessing.cpp (hunk 4 above)
Comment (kaikawaliu) [reply src/libDirectoryService/FinalBlockPreProcessing.cpp]: Added

File: src/libNode/MicroBlockPreProcessing.cpp [hunk 5]
```diff
@@ -1064,6 +1059,25 @@ unsigned char Node::CheckLegitimacyOfTxnHashes(bytes& errorMsg) {
   return LEGITIMACYRESULT::SUCCESS;
 }
 
+bool Node::CheckMicroBlockGasLimit(const uint64_t& microblock_gas_limit) {
+  if (LOOKUP_NODE_MODE) {
+    LOG_GENERAL(WARNING,
+                "Node::CheckMicroBlockHashes not expected to be called "
```
Comment (ansnunez) [inline src/libNode/MicroBlockPreProcessing.cpp]: Wrong log message

File: src/libNode/MicroBlockPreProcessing.cpp [hunk 6]
```diff
@@ -1064,6 +1059,25 @@ unsigned char Node::CheckLegitimacyOfTxnHashes(bytes& errorMsg) {
...
+                "from LookUp node");
+    return true;
+  }
+
+  if (microblock_gas_limit != m_microblock->GetHeader().GetGasLimit()) {
+    LOG_GENERAL(WARNING, "Gas limit check failed. expected: "
```
Comment (ansnunez) [inline src/libNode/MicroBlockPreProcessing.cpp]: Can use `LOG_CHECK_FAIL("Gas limit", m_microblock->GetHeader().GetGasLimit(), microblock_gas_limit)`

File: src/libNode/MicroBlockPreProcessing.cpp (hunk 5 above)
Comment (kaikawaliu) [reply src/libNode/MicroBlockPreProcessing.cpp]: Modified

File: src/libNode/MicroBlockPreProcessing.cpp (hunk 6 above)
Comment (kaikawaliu) [reply src/libNode/MicroBlockPreProcessing.cpp]: Modified

**Pull request merged**